rpm/files.go

```diff
@@ -0,0 +1,139 @@
+package rpm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"sync"
+
+	"github.com/quay/claircore"
+	"github.com/quay/zlog"
+	"golang.org/x/sync/singleflight"
+)
+
+// filesCache is used for concurrent access to the map containing
+// [claircore.Layer] -> map[string]struct{}.
+type filesCache struct {
+	c    map[string]map[string]struct{}
+	refs map[string]int
+	mu   sync.Mutex
+	sf   singleflight.Group
+}
+
+var fc = &filesCache{
+	c:    map[string]map[string]struct{}{},
+	refs: map[string]int{},
+}
+
+// GC decrements the reference counter and deletes the [claircore.Layer]'s
+// entry from the cache map and the reference counter once the references
+// are 0 (or less).
+func (fc *filesCache) gc(key string) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]--
+	if fc.refs[key] <= 0 {
+		delete(fc.c, key)
+		delete(fc.refs, key)
```

Comment:
Should we also garbage collect the `singleflight.Group` memoization? Maybe something like:
```suggestion
		delete(fc.c, key)
		delete(fc.refs, key)
		s.sf.Forget(key)
```

Reply:
This would be an odd situation where the DB extraction was still in progress but the refs were 0 but I guess if there is some unexpected context cancellation then there's a chance that key still exists.

Reply:
Do we ever `Forget()` (or otherwise clear) the results cached in `fc.sf`? My current understanding is that `fc.sf` and `fc.c` contain essentially the same data.

Reply:
`fc.sf` isn't caching results, it's suppressing concurrent executions of the same function by blocking the "non-first" calls until the first call finishes and then sharing the results. If another call is made after the initial calls have finished the function will be run again. i.e Singleflight is implicitly "forgetting".

Reply:
@BradLugo the internal `doCall` has a line `delete(g.m, key)` which is essentially the same as `(*singleflight.Group).Forget(key)`, so it's not needed - https://cs.opensource.google/go/x/sync/+/refs/tags/v0.11.0:singleflight/singleflight.go;l=160

Then again, I guess putting it here won't hurt anyone...? I'm ok either way I think

Reply:
Yeah, I'll stick it in for old time's sake

Reply:
Then again, I think it is possible a single instance of Clair/Claircore may analyze the same layer at the same time, so calling this `Forget` may somehow cause us to redo some of this work. Fun thinking about concurrency 😂

Reply:
dang, this `singleflight` package is cool. Thanks for the explanation yall! 😁